src/providers/postgres/qgspostgresfeatureiterator.cpp

```diff
@@ -990,6 +990,20 @@ void QgsPostgresFeatureIterator::getFeatureAttribute( int idx, QgsPostgresResult
       }
       break;
     }
+    case QVariant::UserType:
+    {
+      if ( fld.type() == QVariant::fromValue( QgsInterval() ).type() && fld.length() == 256 )
```

Comment:
The check `fld.type()` is redundant here it is always `QVariant::UserType`, so basically we are checking for a magic value of `256` to decide if this is an interval or a geometry, which is very loose ground. Can we use something better?

Reply:
What I could imagine
- replace all those `QVariant::fromValue( QgsInterval() ).type()` calls with `QVariant::UserType`.
- drop the `if` statement here
- handle this based on the subtype in `QgsPostgresProvider::convertValue`, like it's already done in the json / hstore case

Reply:
The first part of the if statement, I totally agree. 
For the second part, I believe it is what I originally did. And it worked ... For a while. Then I think one of your PR got merged and I realized that qvariant::user type always have a type Id 1024 : qgsinterval and geometry in particular (this is wierd because qt doc says user type should have id's starting 256 and incrementing....). The test you introduce several days ago ended up getting interval type where geometry was required and thus was failing.
But you are right, introducing a magic number might not be the best solution either (though I do not have any better right now).

Reply:
Do you suggest implementing something here : https://github.com/qgis/QGIS/blob/master/src/providers/postgres/qgspostgresprovider.cpp line 5051? 
Both geometry and QGSInterval are Usertype, with the same ID ... I only saw length that I could leverage to differentiate them. 
Any thoughts?

Reply:
What about the type name?

https://github.com/qgis/QGIS/blob/master/src/providers/postgres/qgspostgresprovider.cpp#L5034